### ISSUE TITLE
Dired-X is not being loaded and causing an error in the vinegar layer

### ISSUE THREAD

**Issue by caleb** [syl20bnr/spacemacs#5705]
#### Description
With a recent change to develop (https://github.com/syl20bnr/spacemacs/pull/5597) `dired-x` is not being loaded for the vinegar layer, and its functions are not defined.

Specifically, if you enable the vinegar layer you get an error that `dired-omit-mode` is not a function but is instead void.

If I add the `(require 'dired-x)` line removed in the mentioned patch back into the `vinegar` layer's keybindings then everything works as expected.

#### Reproduction guide
- Enable the `vinegar` layer
- Start Emacs
- Try to use dired (e.g. hit `-`)

*Observed behaviour:*
Dired does not work.

*Expected behaviour:*
Dired should work :)

#### System Info
- OS: darwin
- Emacs: 24.5.1
- Spacemacs: 0.105.16
- Spacemacs branch: develop (rev. 09ff77f)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: ivy
- Layers:
```elisp
(auto-completion better-defaults c-c   clojure colors dash dockerfile elixir elm emacs-lisp evil-commentary evil-cleverparens evil-snipe eyebrowse git github go html java javascript markdown org osx php python react ruby ruby-on-rails rust
                 (shell :variables shell-default-height 30 shell-default-position 'bottom)
                 search-engine shell-scripts spacemacs-ivy spacemacs-layouts spell-checking sql syntax-checking themes-megapack theming typescript unimpaired version-control vinegar yaml)
```

Thanks,
Caleb

**Comment by darkfeline** [comment]
Can you turn on debugging and see which undefined functions are being called?

If they're being called in Spacemacs we should add that require back in or define the proper autoloads, but if they're being called by vinegar it should be fixed upstream.

**Comment by caleb** [comment]
The function that is missing is `dried-omit-mode`, which vinegar calls directly.

**Comment by StreakyCobra** [comment]
This function should probably be added as a `:commands` in `(use-package dired-x …` then.

**Comment by caleb** [comment]
I managed to get it working by adding an `autoload` for the `dired-omit-mode` function to the `spacemacs-base/init-dired-x`:

```lisp
(defun spacemacs-base/init-dired-x ()
  (autoload 'dired-omit-mode "dired-x"
    "Omit some uninteresting files (backup files, autosave files, etc) from dired listings." t)

  (autoload 'dired-jump "dired-x"
    "Jump to Dired buffer corresponding to current buffer." t)

  (autoload 'dired-jump-other-window "dired-x"
    "Like \\[dired-jump] (dired-jump) but in other window." t))
```

**Comment by StreakyCobra** [comment]
Can you try to rpelace this function by this one:

```elisp 
(defun spacemacs-base/init-dired-x ()
  (use-package dired-x
    :commands (dired-jump
               dired-jump-other-window
               dired-omit-mode)))
```

and check if it works?

**Comment by caleb** [comment]
Using that version does seem to work.

**Comment by caleb** [comment]
The only issue I have now is that in my `dotspacemacs/user-config` I am modifying the list of extensions to hide in `dired` mode:

```lisp
(setq dired-omit-extensions (remove ".mo" dired-omit-extensions))
```

And at this point the `dired-x` extension isn't loaded and the `dried-omit-extensions` variable isn't set, so I get an error.

I can fix it if I just make sure to require dired-x:

```lisp
;; Show .mo files in dired
(require 'dired-x)
(setq dired-omit-extensions (remove ".mo" dired-omit-extensions))
```

**Comment by StreakyCobra** [comment]
Spacemacs use lazy loading of packages so this is expected. The correct way to do it is to use this:

```elisp
(with-eval-after-load 'dired-x
  (setq dired-omit-extensions (remove ".mo" dired-omit-extensions)))
```

**Comment by caleb** [comment]
Thanks! I'm sure this will come in handy for other configurations as well.

**Comment by StreakyCobra** [comment]
This has been fixed in f651969d12c915f24ede4a575a340569f2d07595. I also changed the way of defining the autoloads. If everything is working you can close the issue :-)
